107-tee-boot-attestation.md

```diff
@@ -0,0 +1,360 @@
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+- [ ] Enhancement issue in release milestone, which links to pull request in [keylime/enhancements]
+- [ ] Core members have approved the issue with the label `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place
+- [ ] User-facing documentation has been created in [keylime/keylime-docs]
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+## Summary
+
+<!--
+This section is incredibly important for producing high quality user-focused
+documentation such as release notes or a development roadmap.  It should be
+possible to collect this information before implementation begins in order to
+avoid requiring implementers to split their attention between writing release
+notes and implementing the feature itself. Reviewers
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+-->
+
+Trusted Execution Environments (TEEs) are a confidential computing and
+virtualization technology that allow for the protection/encryption of guest VM
+RAM, cache memory, and CPU registers for running sensitive applications on
+potentially untrusted hosts within the cloud/edge. With TEEs, data written to
+RAM, caches, or registers are encrypted with keys maintained by secure
+processors embedded within a CPU, thus protecting confidential VM guests from
+other guests on a host, or even the host system itself. With encryption fully
+managed by the hardware, only the confidential VM (CVM) itself would be able to
+read/write to its own memory, protecting against buggy/malicious hosts spying
+on or tampering with the CVM.
+
+TEEs are for protecting applications/users on untrusted platforms. In-fact, the
+confidential computing threat model assumes that the host in which a CVM is
+running on top of is actively attempting to tamper or spy on guest memory. With
+this, users must *ensure* that a host has launched their guest with all proper
+TEE protections and didn't perform any other nefarious actions to tamper with
+a CVM. That is, CVMs must *attest* their boot environment and prove the
+*integrity* of their workload before any sensitive operations can be performed
+within a CVM.
+
+This enhancement will introduce a new handler to the keylime verifier strictly
+for performing TEE boot attestation. However, with untrusted hosts controlling
+guest memory, certain attacks can take place *before* a guest OS is booted and a
+keylime agent has been initialized. Timing of boot attestation is critical,
+and thus a TEE boot attestation handler cannot assume that it is communicating
+with a keylime agent. Furthermore, the TEE handler will release some secret
```

Comment:
This would **significantly** change the `keylime` design and mode of operation. The `verifier` is never expected to **block** operations on nodes (e.g., especially one as crucial as the initial boot). `Keylime` has a "scale-out" architecture where `verifiers` can be replicated and are considered mostly stateless (for instance, this is how `Keylime` is deployed in `OpenShift`/`Kubernetes`).